src/document.rs

```diff
@@ -129,172 +172,447 @@ pub struct Prescription<'a> {
+        Ok(MatchType::None)
     }
-
-    Ok(())
 }
 
-/// Extract and transform the element-level Rx prompts from the document content into
-/// contextualized metadata on the match types of those nodes.
-fn process_child_prompts(node: &Node, document: &Document) -> HowserResult<()> {
+fn extract_block_quote_match_token(node: &Node) -> HowserResult<MatchType> {
+    if let Some(paragraph) = node.capabilities
+        .traverse
+        .as_ref()
+        .ok_or(HowserError::CapabilityError)?
+        .first_child()?
     {
-        let traverser = node.capabilities
-            .traverse
+        if paragraph
+            .capabilities
+            .get
             .as_ref()
-            .ok_or(HowserError::CapabilityError)?;
-        let mut child = traverser.first_child()?;
-
-        while let Some(sibling) = child {
-            let sibling_trav = sibling
+            .ok_or(HowserError::CapabilityError)?
+            .get_type()? == NodeType::CMarkNodeParagraph
+        {
+            let token = extract_paragraph_match_token(&paragraph);
+            if paragraph
                 .capabilities
                 .traverse
                 .as_ref()
-                .ok_or(HowserError::CapabilityError)?;
-
-            let mut node_type: NodeType;
-            if let Some(ref getter) = sibling.capabilities.get {
-                node_type = getter.get_type()?;
-            } else {
-                return Err(HowserError::CapabilityError);
-            }
-
-            if helpers::node_type_is_sequencable(&node_type)
-                && !(node_type == NodeType::CMarkNodeList)
+                .ok_or(HowserError::CapabilityError)?
+                .first_child()?
+                .is_none()
             {
-                process_match_token(&sibling, &document)?;
+                paragraph
+                    .capabilities
+                    .mutate
+                    .as_ref()
+                    .ok_or(HowserError::CapabilityError)?
+                    .unlink();
             }
+            token
+        } else {
+            Ok(MatchType::None)
+        }
+    } else {
+        Ok(MatchType::None)
+    }
+}
 
-            process_child_prompts(&sibling, document)?;
+fn extract_code_block_match_token(node: &Node) -> HowserResult<MatchType> {
+    let info = node.capabilities
+        .get
+        .as_ref()
+        .ok_or(HowserError::CapabilityError)?
+        .get_fence_info()?;
+    let (match_type, content) = extract_match_type(&info)?;
+    if match_type != MatchType::None {
+        node.capabilities
+            .set
+            .as_ref()
+            .ok_or(HowserError::CapabilityError)?
+            .set_fence_info(&content)?;
+        Ok(match_type)
+    } else {
+        Ok(MatchType::None)
+    }
+}
 
-            if helpers::node_type_is_circumstantial(&node_type) {
-                process_circumstantial_node(&sibling, document)?;
+fn extract_heading_match_token(node: &Node) -> HowserResult<MatchType> {
```

Comment:
I see a lot of functions in here that take a node, then match on its type, ultimately. This says to me that it would be more naturally represented by using the language's type-based dispatch system, i.e. define one or more traits. Maybe there's a HowerNode trait with all of these?

Reply:
(This is not obviously the right thing to do. But it's something to consider.)

Reply:
Yes, I think there is definitely some abstraction that could be taken advantage of to reduce some type checking. Would you feel ok about me targeting that as an improvement and tracking it as an issue?